samples/boards/st/uart/circular_dma/src/main.c

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2024 STMicroelectronics
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/sys/ring_buffer.h>
+#include <zephyr/logging/log.h>
+#include <string.h>
+
+LOG_MODULE_REGISTER(circular_dma);
+
+#define RING_BUF_SIZE 100
+#define RX_BUF_SIZE   10
+
+#define RECEIVE_TIMEOUT 0
+
+#define STACK_SIZE 1024
+
+#define UART_DEVICE_NODE DT_CHOSEN(zephyr_shell_uart)
+
+static const struct device *const uart_dev = DEVICE_DT_GET(UART_DEVICE_NODE);
+
+/* uart configuration structure */
+const struct uart_config uart_cfg = {.baudrate = 115200,
+				     .parity = UART_CFG_PARITY_NONE,
+				     .stop_bits = UART_CFG_STOP_BITS_1,
+				     .data_bits = UART_CFG_DATA_BITS_8,
+				     .flow_ctrl = UART_CFG_FLOW_CTRL_NONE};
+
+/* define a ring buffer to get raw bytes*/
+RING_BUF_DECLARE(ring_buf, RING_BUF_SIZE);
+
+/* define uart rx buffer */
+static uint8_t rx_buffer[RX_BUF_SIZE];
+
+/* define thread  stack size */
+static K_THREAD_STACK_DEFINE(uart_rx_stack, STACK_SIZE);
+
+/* struct uart_event async_evt */
+static struct k_thread uart_rx_thread_data = {0};
+
+void print_uart(char *buf, int len)
+{
+	for (int i = 0; i < len; i++) {
+		uart_poll_out(uart_dev, buf[i]);
+	}
+}
+
+/* Data processing thread */
+static void uart_rx_thread(void *p1, void *p2, void *p3)
+{
+	uint8_t rx_data[RX_BUF_SIZE];
+	size_t len;
+
+	while (1) {
+
+		/* Check if there's data in the ring buffer */
+		len = ring_buf_get(&ring_buf, rx_data, sizeof(rx_data));
+
+		if (len > 0) {
+
+			/* Process `len` bytes of data */
+			print_uart(rx_data, len);
+			if (ring_buf.get_head == ring_buf.size) {
+				printk("\nring buffer overflow\n");
+			}
+		}
```

Comment:
I'm still puzzled by the fact we're not able to use a blocking mechanism here.

Reply:
I found what to do to make it work with a flag. I update the code